Source/Core/Core/State.cpp

```diff
@@ -525,13 +493,69 @@ void SaveAs(const std::string& filename, bool wait)
       true);
 }
 
+static bool GetVersionFromLZO(StateHeader& header, File::IOFile& f)
+{
+  // Just read the first block, since it will contain the full revision string
+  lzo_uint32 cur_len = 0;  // size of compressed bytes
+  lzo_uint new_len = 0;    // size of uncompressed bytes
+  std::vector<u8> buffer;
+  buffer.resize(header.must_be_zero);
+
+  if (!f.ReadArray(&cur_len, 1) || !f.ReadBytes(out, cur_len))
+    return false;
+
+  const int res = lzo1x_decompress(out, cur_len, buffer.data(), &new_len, nullptr);
+  if (res != LZO_E_OK)
+  {
+    // This doesn't seem to happen anymore.
+    PanicAlertFmtT("Internal LZO Error - decompression failed ({0}) ({2}) \n"
+                   "Unable to retrieve outdated savestate version info.",
+                   res, new_len);
+    return false;
+  }
+
+  // Read in cookie, string length, and the string
+  memcpy(&header.version_cookie, buffer.data(), 2 * sizeof(u32));
+  header.version_str = std::make_unique<char[]>(header.version_str_len);
+  memcpy(header.version_str.get(), buffer.data() + 2 * sizeof(u32), header.version_str_len);
+
+  return true;
+}
+
+static bool ReadStateHeaderFromFile(StateHeader& header, File::IOFile& f)
+{
+  u32 cookie_offset = offsetof(StateHeader, version_cookie);
+
+  if (!f.ReadBytes(&header, cookie_offset))
+  {
+    Core::DisplayMessage("State not found", 2000);
```

Comment:
There's probably some improvements to be made here regarding error reporting. 'no file with that name exists' is a different scenario than 'there is a file here but we don't recognize it as a valid savestate'.